SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift

```diff
@@ -0,0 +1,250 @@
+//
+//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+
+@objc(OWSContactDiscoveryOperation)
+class ContactDiscoveryOperation: OWSOperation {
+
+    let batchSize = 2048
+    let recipientIdsToLookup: [String]
+
+    @objc
+    var registeredRecipientIds: Set<String>
+
+    @objc
+    required init(recipientIdsToLookup: [String]) {
+        self.recipientIdsToLookup = recipientIdsToLookup
+        self.registeredRecipientIds = Set()
+
+        super.init()
+
+        Logger.debug("\(logTag) in \(#function) with recipientIdsToLookup: \(recipientIdsToLookup.count)")
+        for batchIds in recipientIdsToLookup.chunked(by: batchSize) {
+            let batchOperation = LegacyContactDiscoveryBatchOperation(recipientIdsToLookup: batchIds)
+            self.addDependency(batchOperation)
```

Comment:
So, if any of the operations in the batch fail, we treat the entire batch as a failure, right?  So if any fail, shouldn't we cancel any outstanding operations?

Also, `OWSOperation` does retry, but without exponential backoff.  Don't we need to change that now?

Reply:
OWSOperation **can** do retry, but only if you set `remainingRetries` to something greater than 0.

For now, none of these options retry, and I've left all retry logic in the callers (OWSContactManager#intersectContacts:retryDelaySexconds...)

Reply:
> So, if any of the operations in the batch fail, we treat the entire batch as a failure, right? So if any fail, shouldn't we cancel any outstanding operations?

Yeah, good idea. I'll make this change.

Reply:
I implemented the dependent cancel logic here:

https://github.com/signalapp/Signal-iOS/pull/3726/commits/2e30b33415a3cad161632f9f63f26c467c5f1c87